dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStageDataElementStore.java

```diff
@@ -88,4 +92,28 @@ public List<ProgramStageDataElement> getAllByDataElement(Collection<DataElement>
         .setParameter("dataElements", dataElements)
         .list();
   }
+
+  @Override
+  public Map<String, Set<String>> getProgramStageDataElementsWithSkipSynchronizationSetToTrue() {
+    final String sql =
+        "select ps.uid as ps_uid, de.uid as de_uid from programstagedataelement psde "
+            + "join programstage ps on psde.programstageid = ps.programstageid "
+            + "join dataelement de on psde.dataelementid = de.dataelementid "
+            + "where psde.programstageid in (select distinct ( programstageid ) from event ev where ev.lastupdated > ev.lastsynchronized) "
```

Comment:
Why are we checking event table here at all?

Reply:
I think we are overoptimizing with a very higher cost here. 

If we get ALL PSDEs for all program stages (But within a single program), the query will be very quick as we are dealing with metadata tables along. 

But in this case, If we get PSDE lookup that only actually have unsynchronized events, we are actually touching event table which is potentially huge. This query could be slower or heavier.

If we get all PSDEs for all program stages (quickly) and then later when fetching unsyncrhonized events do the filtering, we would still consume the same cost (but a single time). 

So my question is, what will be the changes if we get all PSDEs for all program stages in that program, irrespective of unsynchronized events.